### PR TITLE
[LETS-340] transaction server server-server error handling

### PR DESCRIPTION
**Pull request by cristiarg**
http://jira.cubrid.org/browse/LETS-340

Server-server connection error handling for different call stacks that transaction server uses.
   
Implementation:
- callstack `init_server_type` - `tran_server::boot` -> `tran_server::get_boot_info_from_page_server` -> `tran_server::send_receive` which will result in the active transactions server failing to boot
- callstack `pgbuf_fix` -> `pgbuf_claim_bcb_for_fix` -> `pgbuf_read_page_from_file_or_page_server` -> `pgbuf_request_data_page_from_page_server` -> `tran_server::send_receive` which will result in failing to fix the heap page
- callstack `logpb_read_page_from_file_or_page_server` -> `logpb_request_log_page_from_page_server` -> `tran_server::send_receive` which will fail in fixing the log page
- callstack `log_initialize_passive_tran_server` -> `passive_tran_server::send_and_receive_log_boot_info` -> `tran_server::send_receive` which will result in passive transaction server failing to boot
- in `passive_tran_server::send_and_receive_stop_log_prior_dispatch` error handling is not implemented because passive transaction server is going down anyway

### CODE REVIEW

File: src/communication/response_broker.hpp [hunk 1]
```diff
@@ -82,76 +105,112 @@ namespace cubcomm
...
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::get_response (response_sequence_number a_rsn)
   {
     std::unique_lock<std::mutex> ulock (m_mutex);
-    T_PAYLOAD payload;
+    std::tuple<T_PAYLOAD, T_ERROR> payload_or_error;
 
-    auto condvar_pred = [this, &payload, a_rsn] ()
+    auto condvar_pred = [this, &payload_or_error, a_rsn] ()
     {
       auto it = m_response_payloads.find (a_rsn);
```
Comment (alexstanro) [inline src/communication/response_broker.hpp]: What's happen if a_rsn is not found?

File: src/communication/request_sync_send_queue.hpp [hunk 2]
```diff
@@ -169,14 +171,15 @@ namespace cubcomm
...
-      payload_type &&payload)
+      payload_type &&payload, send_queue_error_handler &&error_handler)
   {
     // synchronize push request into the queue and notify consumers
 
     std::unique_lock<std::mutex> ulock (m_queue_mutex);
     m_request_queue.emplace ();
     m_request_queue.back ().m_id = reqid;
     m_request_queue.back ().m_payload = std::move (payload);
+    m_request_queue.back ().m_error_handler = std::move (error_handler);
```
Comment (alexstanro) [inline src/communication/request_sync_send_queue.hpp]: Only one error_handler is registered for request_sync_send_queue. Do we need to add error handler for each element in this queue?

File: src/communication/response_broker.hpp [hunk 3]
```diff
@@ -82,76 +105,112 @@ namespace cubcomm
...
+    }
+    m_condvar.notify_all (); // all because there is more than one thread waiting for data
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  void
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::register_error (response_sequence_number a_rsn, T_ERROR &&a_error)
+  {
+    {
+      std::lock_guard<std::mutex> lockg (m_mutex);
```
Comment (alexstanro) [inline src/communication/response_broker.hpp]: Be careful, this mutex may be acquired while holding m_send_mutex.

**Review by alexstanro — APPROVED** [review]

File: src/communication/request_sync_send_queue.hpp (hunk 2 above)
Comment (cristiarg) [reply src/communication/request_sync_send_queue.hpp]: Outdated code belonging to other patch.

File: src/communication/response_broker.hpp (hunk 3 above)
Comment (cristiarg) [reply src/communication/response_broker.hpp]: Outdated code belonging to other patch.

File: src/communication/response_broker.hpp (hunk 1 above)
Comment (cristiarg) [reply src/communication/response_broker.hpp]: Outdated code belonging to other patch.

**Pull request merged**